transport/src/main/java/io/netty/bootstrap/ChannelInitializerExtensions.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2023 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.bootstrap;
+
+import io.netty.util.internal.SystemPropertyUtil;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.ServiceLoader;
+
+/**
+ * The configurable facade that decides what {@link ChannelInitializerExtension}s to load and where to find them.
+ */
+abstract class ChannelInitializerExtensions {
+    private static volatile ChannelInitializerExtensions implementation;
+
+    private ChannelInitializerExtensions() {
+    }
+
+    /**
+     * Get the implementation where we have no extensions.
+     */
+    static ChannelInitializerExtensions empty() {
+        return EmptyExtensions.EMPTY;
+    }
+
+    /**
+     * Get the configuration extensions, which may be the same as {@link #empty()} if the
+     * {@code io.netty.bootstrap.extensions} system property is {@code empty} or {@code false}.
+     * <p>
+     * By default, we pick an implementation that finds extensions through
+     * {@linkplain ServiceLoader#load(Class) service loading}.
+     */
+    static ChannelInitializerExtensions getExtensions() {
+        ChannelInitializerExtensions impl = implementation;
+        if (impl == null) {
+            synchronized (ChannelInitializerExtensions.class) {
+                impl = implementation;
+                if (impl != null) {
+                    return impl;
+                }
+                String extensionProp = SystemPropertyUtil.get("io.netty.bootstrap.extensions");
```

Comment:
Why enable by default? Big part of netty 4.1 users unlikely to use this, and service loaders are not free at runtime.

Reply:
The service loading is only done once, when the first channel is initialized. Then the list is cached.

Still, flipping the default is probably a good idea. Service loading is a bit of magic behavior, and having magic be opt-in at application launch time sounds like the appropriate way to get consent.

Reply:
Not a strong opinion, but I would probably skip adding a system property too until we have a use-case.

Today, there are no extensions - nothing to worry about.
If someone adds an extension, they know what they are doing and extension is expected to work. 
If an application owner wants to remove any extension, they can modify resulting `META-INF/services` when they package application to exclude that specific extension.

In regards to loading, consider loading extensions once when `AbstractBootstrap` is initialized instead of affecting runtime of the first channel creation.

Reply:
` skip adding a system property too until we have a use-case.` - extensions have to be disabled by default, since they are essentially a backdoor - any 3rd party code may declare service (either classpath META-INF/services or java module) as extension, and netty will start to blindly apply them to all otherwise private channels - without explicit permission to do so - which is surprising at least, and in fact is major security concern.

Regarding `owner wants to remove any extension` - `META-INF/services` is part of the (3rd party dependency) jar so It is not easy to remove them.

Reply:
I agree with @mostroverkhov on this one.

Reply:
Good points @mostroverkhov, let's keep an opt-in system property.
Do you think a "boolean" property is enough or should it be a list of acceptable extensions as a full class name?

Reply:
@idelpivnitskiy IMO having boolean property seems enough until more extension sources are added - then source name may be provided with additional string property (but I postpone decision on this to actual users of the new API).